sky/data/storage.py

```diff
@@ -843,47 +881,91 @@ def delete(self) -> None:
         """
-        source = os.path.abspath(os.path.expanduser(self.source))
-        sync_command = ('aws s3 sync --no-follow-symlinks '
-                        f'{source} s3://{self.name}/')
+        # Generate message for upload
+        if len(source_path_list) > 1:
+            source_message = f'{len(source_path_list)} paths'
+        else:
+            source_message = source_path_list[0]
+
+        # Generate gsutil rsync command for files and dirs
+        commands = []
+        grouped_files, dirs = data_utils.group_files_by_dir(source_path_list)
+        # Generate file upload commands
+        for dir_path, file_names in grouped_files.items():
+            includes = ' '.join(
+                [f'--include "{file_name}"' for file_name in file_names])
+            sync_command = ('aws s3 sync --no-follow-symlinks --exclude="*" '
+                            f'{includes} {dir_path} '
+                            f's3://{self.name}')
+            commands.append(sync_command)
+        # Generate dir upload commands
+        for dir_path in dirs:
+            if create_dirs:
+                dest_dir_name = os.path.basename(dir_path)
+            else:
+                dest_dir_name = ''
+            sync_command = ('aws s3 sync --no-follow-symlinks '
+                            f'{dir_path} '
+                            f's3://{self.name}/{dest_dir_name}')
+            commands.append(sync_command)
+
+        # Run commands in parallel
         with backend_utils.safe_console_status(
                 f'[bold cyan]Syncing '
-                f'[green]{self.source}[/] to [green]s3://{self.name}/[/]'):
-            # TODO(zhwu): Use log_lib.run_with_log() and redirect the output
-            # to a log file.
-            with subprocess.Popen(sync_command.split(' '),
-                                  stderr=subprocess.PIPE,
-                                  stdout=subprocess.DEVNULL) as process:
-                stderr = []
-                while True:
-                    line = process.stderr.readline()
-                    if not line:
-                        break
-                    str_line = line.decode('utf-8')
-                    stderr.append(str_line)
-                    if 'Access Denied' in str_line:
-                        process.kill()
-                        with ux_utils.print_exception_no_traceback():
-                            raise PermissionError(
-                                'Failed to upload files to '
-                                'the S3 bucket. The bucket does not have '
-                                'write permissions. It is possible that '
-                                'the bucket is public.')
-                returncode = process.wait()
-                if returncode != 0:
-                    stderr = '\n'.join(stderr)
+                f'[green]{source_message}[/] to [green]s3://{self.name}/[/]'):
+            with pool.ThreadPool(processes=_MAX_CONCURRENT_UPLOADS) as p:
```

Comment:
We should use existing code in backend.py that can run several commands in parallel: https://github.com/skypilot-org/skypilot/blob/2336d75fc21d04329634b89d0424d6dceaad698b/sky/backends/cloud_vm_ray_backend.py#L2041

Reply:
In the refactor, it's not possible to reuse because this code chunk now requires using pool.starmap and a custom number of workers

Reply:
Will this correctly fail the other threads when one child fails? Just want to make sure that there will be no leaked thread.
Here is the reason we switched to `p.imap` in `subprocess_utils`.
https://stackoverflow.com/questions/25790279/python-multiprocessing-early-termination

Reply:
@Michaelvll - yes, seems like if an exception is raised, `pool.starmap` correctly terminates the ThreadPool and threads don't leak. The error in the stackoverflow link happens only if `sys.exit` is called in the run function. I tested with this snippet:

```
import subprocess
import sys
from multiprocessing import pool


def run_fn(command, message):
    print(message)
    if '1' in command:
        command += '; asdf'  # Inject error
    with subprocess.Popen(command,
                          stderr=subprocess.PIPE,
                          stdout=subprocess.DEVNULL,
                          shell=True) as process:
        stderr = []
        while True:
            line = process.stderr.readline()
            if not line:
                break
            str_line = line.decode('utf-8')
            stderr.append(str_line)
        returncode = process.wait()
        if returncode != 0:
            # sys.exit(0) # p.map or starmap is problematic only if you raise sys.exit here
            raise RuntimeError(f'Command {command} failed.')


commands = [[f"echo {i}", f"Command{i}"] for i in range(0, 10)]
with pool.ThreadPool(processes=16) as p:
    res = p.starmap(run_fn, commands)
```